src/prefect/new_flow_engine.py

```diff
@@ -224,25 +279,22 @@ async def run_flow(
         # This is a context manager that keeps track of the state of the flow run.
         await run.begin_run()
 
-        while run.is_pending():
-            await asyncio.sleep(1)
-            await run.begin_run()
-
         while run.is_running():
-            try:
-                # This is where the flow is actually run.
-                if flow.isasync:
-                    result = cast(R, await flow.fn(**(run.parameters or {})))  # type: ignore
-                else:
-                    result = cast(R, flow.fn(**(run.parameters or {})))  # type: ignore
-                # If the flow run is successful, finalize it.
-                await run.handle_success(result)
-                if return_type == "result":
-                    return result
-
-            except Exception as exc:
-                # If the flow fails, and we have retries left, set the flow to retrying.
-                await run.handle_exception(exc)
+            async with run.enter_run_context():
+                try:
+                    # This is where the flow is actually run.
+                    if flow.isasync:
+                        result = cast(R, await flow.fn(**(run.parameters or {})))  # type: ignore
+                    else:
+                        result = cast(R, flow.fn(**(run.parameters or {})))  # type: ignore
+                    # If the flow run is successful, finalize it.
+                    await run.handle_success(result)
+                    if return_type == "result":
+                        return result
```

Comment:
is this return statement needed here for any reason?

Reply:
I don't think so. I think we're good with the `state.result()` call at the end of `run_flow`.